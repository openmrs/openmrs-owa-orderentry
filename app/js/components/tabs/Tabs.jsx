import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Tabs extends Component {
    state = {
      activeTabIndex: this.props.defaultActiveTabIndex,
    };

    handleTabClick = (tabIndex, tabName) => {
      this.setState({
        activeTabIndex: tabIndex,
      });
      this.props.careSetting(tabName);
    }

    passPropsToChildren = () => (
      React.Children.map(this.props.children, (child, index) => React.cloneElement(child, {
        onClick: this.handleTabClick,
        tabIndex: index,
        isActive: index === this.state.activeTabIndex,
      })));

    renderActiveTabChildren = () => {
      const { children } = this.props;
      const { activeTabIndex } = this.state;
      if (children[activeTabIndex]) {
        return children[activeTabIndex].props.children;
      }
    }

    render() {
      return (
        <div className="ui-tabs">
          <ul className="ui-tabs-nav ui-helper-reset ui-helper-clearfix ui-widget-header ui-corner-all" role="tablist">
            { this.passPropsToChildren() }
          </ul>
          <div className="ui-tabs-panel ui-widget-content ui-corner-bottom">
            { this.renderActiveTabChildren() }
          </div>
        </div>
      );
    }
}
Tabs.propTypes = {
  defaultActiveTabIndex: PropTypes.number,
  children: PropTypes.node.isRequired,
};

Tabs.defaultProps = {
  defaultActiveTabIndex: 0,
};

export default Tabs;
