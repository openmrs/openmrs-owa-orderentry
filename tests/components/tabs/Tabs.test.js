import React from 'react';
import Tabs from '../../../app/js/components/tabs/Tabs';

describe('`Tabs` component', () => {
    it('should render without crashing', () => {
        const props = {
            closeFormsOnTabChange: jest.fn(),
        }
        const wrapper = shallow(<Tabs {...props}><p>hello</p></Tabs>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render contents at index 0 by default', () => {
        const props = {
            closeFormsOnTabChange: jest.fn(),
        }
        const wrapper = mount(<Tabs {...props}>
            <div tabName="tab 1"><p>Content 1</p></div>
            <div tabName="tab 2"><p>Content 2</p></div>
            <div tabname="tab 3"><p>Content 3</p></div>
            </Tabs>);
        expect(wrapper.find('.ui-tabs-panel').text()).toEqual('Content 1');
    });

    it('should render only content of active children', () => {
        const props = {
            closeFormsOnTabChange: jest.fn(),
        }
        const wrapper = mount(<Tabs defaultActiveTabIndex={1} {...props}>
            <div tabName="tab 1"><p>Content 1</p></div>
            <div tabName="tab 2"><p>Content 2</p></div>
            <div tabname="tab 3"><p>Content 3</p></div>
            </Tabs>);
        expect(wrapper.find('.ui-tabs-panel').text()).toEqual('Content 2');
    });

    it('should change the content when active index changes', () => {
        const props = {
            careSetting : jest.fn(),
            closeFormsOnTabChange: jest.fn(),
            clearSearchField: jest.fn(),
        }
        const wrapper = mount(<Tabs defaultActiveTabIndex={1} {...props}>
            <div tabName="tab 1"><p>Content 1</p></div>
            <div tabName="tab 2"><p>Content 2</p></div>
            <div tabname="tab 3"><p>Content 3</p></div>
            </Tabs>);
        const children = (wrapper.find('.ui-tabs-nav').children());
        wrapper.instance().handleTabClick(2, 'tab 3');
        expect(wrapper.find('.ui-tabs-panel').text()).toEqual('Content 3');
    });
});
