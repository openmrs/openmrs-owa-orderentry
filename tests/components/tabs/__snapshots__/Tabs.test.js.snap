// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Tabs\` component should render without crashing 1`] = `
<div
  className="ui-tabs"
>
  <ul
    className="ui-tabs-nav ui-helper-reset ui-helper-clearfix ui-widget-header ui-corner-all"
    role="tablist"
  >
    <p
      isActive={true}
      key=".0"
      onClick={[Function]}
      tabIndex={0}
    >
      hello
    </p>
  </ul>
  <div
    className="ui-tabs-panel ui-widget-content ui-corner-bottom"
  />
</div>
`;
